mockgcp/mockcertificatemanager/v1.go

```diff
@@ -88,3 +89,125 @@ func (s *CertificateManagerV1) DeleteCertificate(ctx context.Context, req *pb.De
 
 	return s.operations.NewLRO(ctx)
 }
+
+func (s *CertificateManagerV1) GetCertificateMap(ctx context.Context, req *pb.GetCertificateMapRequest) (*pb.CertificateMap, error) {
+	name, err := s.parseCertificateMapName(req.Name)
+	if err != nil {
+		return nil, err
+	}
+
+	if time.Now().UnixNano()%2 == 0 {
+		return nil, status.Errorf(codes.Internal, "try again!")
+	}
+
+	fqn := name.String()
+
+	obj := &pb.CertificateMap{}
+	if err := s.storage.Get(ctx, fqn, obj); err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil, status.Errorf(codes.NotFound, "certificateMap %q not found", name)
+		} else {
+			return nil, status.Errorf(codes.Internal, "error reading certificateMap: %v", err)
+		}
+	}
+
+	return obj, nil
+}
+
+func (s *CertificateManagerV1) CreateCertificateMap(ctx context.Context, req *pb.CreateCertificateMapRequest) (*longrunning.Operation, error) {
+	reqName := req.Parent + "/certificateMaps/" + req.CertificateMapId
+	name, err := s.parseCertificateMapName(reqName)
+	if err != nil {
+		return nil, err
+	}
+
+	fqn := name.String()
+
+	obj := proto.Clone(req.CertificateMap).(*pb.CertificateMap)
+	obj.Name = fqn
+
+	if err := s.storage.Create(ctx, fqn, obj); err != nil {
+		return nil, status.Errorf(codes.Internal, "error creating certificate map: %v", err)
+	}
+
+	return s.operations.NewLRO(ctx)
+}
+
+func (s *CertificateManagerV1) DeleteCertificateMap(ctx context.Context, req *pb.DeleteCertificateMapRequest) (*longrunning.Operation, error) {
+	name, err := s.parseCertificateMapName(req.Name)
+	if err != nil {
+		return nil, err
+	}
+
+	fqn := name.String()
+
+	kind := (&pb.CertificateMap{}).ProtoReflect().Descriptor()
+	if err := s.storage.Delete(ctx, kind, fqn); err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil, status.Errorf(codes.NotFound, "certificate map %q not found", name)
+		} else {
+			return nil, status.Errorf(codes.Internal, "error deleting certificate map: %v", err)
+		}
+	}
+
+	return s.operations.NewLRO(ctx)
+}
+
+func (s *CertificateManagerV1) GetDnsAuthorization(ctx context.Context, req *pb.GetDnsAuthorizationRequest) (*pb.DnsAuthorization, error) {
+	name, err := s.parseDNSAuthorizationName(req.Name)
+	if err != nil {
+		return nil, err
+	}
+
+	fqn := name.String()
+
+	obj := &pb.DnsAuthorization{}
+	if err := s.storage.Get(ctx, fqn, obj); err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil, status.Errorf(codes.NotFound, "dns authorization %q not found", name)
+		} else {
+			return nil, status.Errorf(codes.Internal, "error reading dns authorization: %v", err)
+		}
+	}
+
+	return obj, nil
+}
+
+func (s *CertificateManagerV1) CreateDnsAuthorization(ctx context.Context, req *pb.CreateDnsAuthorizationRequest) (*longrunning.Operation, error) {
+	reqName := req.Parent + "/dnsAuthorizations/" + req.DnsAuthorizationId
```

Comment:
For my own education, how is the format for `reqName` determined? I am looking at Terraform implementation, the create URL looks like this `{{CertificateManagerBasePath}}projects/{{project}}/locations/global/dnsAuthorizations?dnsAuthorizationId={{name}}` [1]. I did not look into the actual GCP implementation on how the url is parsed.


[1] https://github.com/GoogleCloudPlatform/k8s-config-connector/blob/44bfcef1694935b669a6b1e72f4fc54f7784eb71/third_party/github.com/hashicorp/terraform-provider-google-beta/google-beta/services/certificatemanager/resource_certificate_manager_dns_authorization.go#L143C49-L143C164

Reply:
So this is probably the hardest bit of the mock :-)  What we're trying to do is establish the "fully qualified name" of the resource.  The only reason we actually need that in most of our mocks is for a key in our map of objects, so we create it / delete it / get it with the same key.  Arguably we could use anything - for example here we could have a type like `type Key struct { Project string; DNSAuthorizationID string; }`.  But I think we will want to populate the selfLink in at least some APIs, so using the same key that GCP uses externally seems to be an easier choice.

In a get or a delete, we normally get this key directly because grpc-gateway parses the URL into a string field in the request object.  In the create, there's a sort of pattern that we get the _parent_ "key", and then we have to add on the child name etc.  My guess is that we have some AIP somewhere that recommends this, as most of our recent APIs seem to follow this pattern!

Reply:
Thank you for the explanation, it makes sense. I was just making sure I did not miss anything.